mllib/src/main/scala/org/apache/spark/ml/feature/FuncTransformer.scala

```diff
@@ -0,0 +1,150 @@
+        s"FuncTransformer outputDataType cannot be automatically inferred, please try" +
+          s" the constructor with specific outputDataType")
+    }
+   )
+
+  setDefault(inputCol -> "input", outputCol -> "output")
+
+  @Since("2.3.0")
+  override def createTransformFunc: IN => OUT = func
+
+  @Since("2.3.0")
+  override def write: MLWriter = new FuncTransformerWriter(
+    this.asInstanceOf[FuncTransformer[Nothing, Nothing]])
+
+  @Since("2.3.0")
+  override def copy(extra: ParamMap): FuncTransformer[IN, OUT] = {
+    copyValues(new FuncTransformer(uid, func, outputDataType), extra)
+  }
+
+  override protected def validateInputType(inputType: DataType): Unit = {
+    try {
+      val funcINType = ScalaReflection.schemaFor[IN].dataType
+      require(inputType.equals(funcINType),
+        s"$uid only accept input type $funcINType but got $inputType.")
+    } catch {
+      case _: UnsupportedOperationException =>
+        // cannot infer the output data type, log warning but do not block transform
+        logWarning(s"FuncTransformer input Type cannot be automatically inferred," +
+          s"Type check omitted for $uid")
+    }
+  }
+}
+
+/**
+ * :: DeveloperApi ::
+ * Companion object for FuncTransformer with save and load function.
+ */
+@DeveloperApi
+@Since("2.3.0")
+object FuncTransformer extends DefaultParamsReadable[FuncTransformer[Nothing, Nothing]] {
+
+  private[FuncTransformer]
+  class FuncTransformerWriter(instance: FuncTransformer[Nothing, Nothing]) extends MLWriter {
+
+    private case class Data(func: Array[Byte], dataType: String)
+
+    override protected def saveImpl(path: String): Unit = {
+      DefaultParamsWriter.saveMetadata(instance, path, sc)
+      val bo = new ByteArrayOutputStream()
+      new ObjectOutputStream(bo).writeObject(instance.func)
+      val data = Data(bo.toByteArray, instance.outputDataType.json)
+      val dataPath = new Path(path, "data").toString
+      sparkSession.createDataFrame(Seq(data)).repartition(1).write.parquet(dataPath)
+    }
+  }
+
+  private class FuncTransformerReader extends MLReader[FuncTransformer[Nothing, Nothing]] {
+
+    private val className = classOf[FuncTransformer[Nothing, Nothing]].getName
+
+    override def load(path: String): FuncTransformer[Nothing, Nothing] = {
+      val metadata = DefaultParamsReader.loadMetadata(path, sc, className)
+      val dataPath = new Path(path, "data").toString
+      val data = sparkSession.read.parquet(dataPath)
+      val Row(funcBytes: Array[Byte], dataType: String) = data
+          .select("func", "dataType")
+          .head()
+      val func = new ObjectInputStream(new ByteArrayInputStream(funcBytes)).readObject()
+      val model = new FuncTransformer(
+        metadata.uid, func.asInstanceOf[Function[Any, Any]], DataType.fromJson(dataType))
+      DefaultParamsReader.getAndSetParams(model, metadata)
+      model.asInstanceOf[FuncTransformer[Nothing, Nothing]]
```

Comment:
After loading, the `FuncTransformer[Nothing, Nothing]` will cause warning during `validateInputType`. The warning will not block `transform`, but it would be better if we can fully restore the `FuncTransformer` with original Type. 

I'll keep checking, but any suggestion is welcome.